pkg/openthread/contrib/netdev/openthread_netdev.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (C) 2017 Fundacion Inria Chile
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ * @ingroup     net
+ * @file
+ * @brief       Netdev adoption for OpenThread
+ *
+ * @author      Jose Ignacio Alamos <jialamos@uc.cl>
+ * @}
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <string.h>
+#include "msg.h"
+#include "openthread/cli.h"
+#include "openthread/instance.h"
+#include "openthread/ip6.h"
+#include "openthread/platform/alarm.h"
+#include "openthread/platform/uart.h"
+#include "openthread/tasklet.h"
+#include "openthread/thread.h"
+#include "random.h"
+#include "ot.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#define OPENTHREAD_QUEUE_LEN      (8)
+static msg_t _queue[OPENTHREAD_QUEUE_LEN];
+
+static kernel_pid_t _pid;
+static otInstance *sInstance;
+
+void ot_exec_job(OT_JOB (*job)(otInstance *, void *), void *context)
+{
+    ot_job_t _job;
+
+    _job.function = job;
+    _job.context = context;
+
+    msg_t msg, reply;
+    msg.type = OPENTHREAD_JOB_MSG_TYPE_EVENT;
+    msg.content.ptr = &_job;
+    msg_send_receive(&msg, &reply, openthread_get_pid());
+}
+
+/* OpenThread will call this when switching state from empty tasklet to non-empty tasklet. */
+void otTaskletsSignalPending(otInstance *aInstance)
+{
+    otTaskletsProcess(aInstance);
+}
+
+static void *_openthread_event_loop(void *arg)
+{
+    _pid = thread_getpid();
+
+    /* enable OpenThread UART */
+    otPlatUartEnable();
+
+    /* init OpenThread */
+    sInstance = otInstanceInit();
+
+    msg_init_queue(_queue, OPENTHREAD_QUEUE_LEN);
+    netdev_t *dev;
+    msg_t msg, reply;
+
+    otCliUartInit(sInstance);
+
+#if OPENTHREAD_ENABLE_DIAG
+    diagInit(sInstance);
+#endif
+
+    uint8_t *buf;
+    ot_job_t *job;
+    while (1) {
+        msg_receive(&msg);
+        switch (msg.type) {
+            case OPENTHREAD_XTIMER_MSG_TYPE_EVENT:
+                /* Tell OpenThread a time event was received */
+                otPlatAlarmFired(sInstance);
+                break;
+            case OPENTHREAD_NETDEV_MSG_TYPE_EVENT:
+                /* Received an event from driver */
+                dev = msg.content.ptr;
+                dev->driver->isr(dev);
+                break;
+            case OPENTHREAD_SERIAL_MSG_TYPE_EVENT:
+                /* Tell OpenThread about the reception of a CLI command */
+                buf = msg.content.ptr;
+                otPlatUartReceived(buf, strlen((char *) buf));
+                break;
+            case OPENTHREAD_JOB_MSG_TYPE_EVENT:
+                job = msg.content.ptr;
+                job->function(sInstance, job->context);
+                msg_reply(&msg, &reply);
+                break;
+        }
+    }
+
+    return NULL;
+}
+
+static void _event_cb(netdev_t *dev, netdev_event_t event)
+{
+    msg_t msg;
```

Comment:
msg should be declared in the `NETDEV_EVENT_ISR` block (use {} to avoid compilation errors)